crates/bevy_render/src/view/visibility/mod.rs

```diff
@@ -70,9 +78,61 @@ impl VisibleEntities {
     }
 }
 
+/// Tracks which [`Entities`](Entity) have which meshes for entities whose [`Aabb`]s are managed by
+/// the [`calculate_bounds`] and [`update_bounds`] systems. This is needed because `update_bounds`
+/// recomputes `Aabb`s for entities whose mesh has been mutated. These mutations are visible via
+/// [`AssetEvent<Mesh>`](AssetEvent) which tells us which mesh was changed but not which entities
+/// have that mesh.
+#[derive(Debug, Default, Clone)]
+pub struct EntityMeshMap {
+    entities_with_mesh: HashMap<Handle<Mesh>, SmallVec<[Entity; 1]>>,
+    mesh_for_entity: HashMap<Entity, Handle<Mesh>>,
+}
+
+impl EntityMeshMap {
+    /// Register the passed `entity` as having the passed `mesh_handle`.
+    fn register(&mut self, entity: Entity, mesh_handle: &Handle<Mesh>) {
+        // Note that this list can have duplicates if an entity is registered for a mesh multiple
+        // times. This should be rare and only cause an additional `Aabb.clone()` in
+        // `update_bounds` so it is preferable to a `HashSet` for now.
+        self.entities_with_mesh
+            .entry(mesh_handle.clone_weak())
+            .or_default()
+            .push(entity);
+        self.mesh_for_entity
+            .insert(entity, mesh_handle.clone_weak());
+    }
+
+    /// Deregisters the mapping between an `Entity` and `Mesh`. Used so [`update_bounds`] can
+    /// track which mappings are still active so `Aabb`s are updated correctly.
+    fn deregister(&mut self, entity: Entity) {
+        let mut inner = || {
+            let mesh = self.mesh_for_entity.remove(&entity)?;
+            let entities = self.entities_with_mesh.get_mut(&mesh)?;
+
+            // There could be duplicate entries in here if an entity was registered with a mesh
+            // multiple times. It's important to remove all references so that if an entity gets a
+            // new mesh and its old mesh is mutated, the entity doesn't get its old mesh's new
+            // `Aabb`.
+            let idxs_to_remove: SmallVec<[usize; 1]> = entities
+                .iter()
+                .enumerate()
+                .filter_map(|(i, e)| (e == &entity).then(|| i))
+                .collect();
+            let num_removed = idxs_to_remove.len();
+            for idx in idxs_to_remove.into_iter().rev() {
+                entities.swap_remove(idx);
+            }
+            Some(num_removed)
```

Comment:
It's minor but I think you could just iterate the indices in reverse and swap_remove if the entity matches.

Reply:
Ah yeah good point!

I did this so I could grab the `len()` but I guess I could just increment a counter? I wanted to record the length because it's an indication of the state of things being "wrong". In my head the actual return type of `deregister` could be:
```rust
enum SomeError {
    EntityInOneMapButNotTheOther,
    EntityRegisteredWithMeshMultipleTimes(usize),
}
fn deregister() -> Option<Result<(), SomeError>> {}
```
Though technically it's probably not useful to know how many times the entity was registered, just that it was more than once.

For this PR do you think I should:
1. leave it as is
2. don't collect and keep track of the number of items so we can return it/log it/something
3. don't collect and keep a bool to see if we removed the entity more than once with a comment indicating something would be wrong in that case
4. don't collect and don't keep track of anything

Reply:
As mentioned, I won’t block on it so 1 for this PR. But a few things to consider for the other answers in a separate PR or this one if you feel like it:

- Is the error a bug in bevy?
- Is it a bug in user code?
- Is the number of items removed necessary to identify that the error occurred?
- Does seeing the number of items removed help with debugging?
- What is simple to implement and understand?

Reply:
Ah no I can't trivially iterate the indices without collecting because the iterator over the indices immutably borrows `entities` which is what I'm trying to remove from. Unless I misunderstood your comment!

Reply:
Oh wait I probably can't do that for the same reason - the reverse iterator over the entities would borrow things when trying to remove?

Reply:
Sorry for the spam, the marbles finally clicked together, I'm guessing you meant something like:
```rust
            for i in (0..entities.len()).rev() {
                if entities[i] == entity {
                    entities.swap_remove(i);
                }
            }
```

Reply:
Updated in the latest commit (0a2ac7c2678931bb8e47b2a0248392e6d7848d11). I also removed the count for now because :shrug: but left some comments indicating to future devs what things might indicate problems.

Reply:
Yup, that was what I meant. :)